FileSpecs/Maps.md

```diff
@@ -0,0 +1,206 @@
+
+* A: Precisely one of {`01`, `02`, `04`} is set on every map
+* B: Precisely one of {`20`, `40`, `80`} is set on every map
+
+
+## 2D Maps
+
+| Offset              | Format             | Name            | Description                    |
+|---------------------|--------------------|-----------------|--------------------------------|
+| ...                 |                    |                 | (cf. Map Header and Structure) |
+| `ac5`               | u8[width * height] | map[0]          | map data                       |
+| `ac5` + mapsize     | u8[width * height] | map[1]          | overlay map                    |
+| `ac5` + mapsize * 2 | u8[width * height] | hotspots        | hotspot map                    |
+| `ac5` + mapsize * 3 | u8[..]             | NPC coordinates | See below                      |
+
+where:
+* mapsize = height * width
+
+
+## 3D maps
+
+| Offset                            | Format             | Name               | Description                    |
+|-----------------------------------|--------------------|--------------------|--------------------------------|
+| ...                               |                    |                    | (cf. Map Header and Structure) |
+| `ac5`                             | u8                 | #labinfo           | Number of LabInfo blocks       |
+| `ac6`                             | u8[4 * #labinfo]   | labinfo[i].head    | See below                      |
+| `ac6` + #labinfo * 4              | u8[#labinfo]       | labinfo[i].rest[0] | See below                      |
+| `ac6` + #labinfo * 5              | u8[#labinfo]       | labinfo[i].rest[1] | See below                      |
+| `ac6` + #labinfo * 6              | u8[#labinfo]       | labinfo[i].rest[2] | See below                      |
+| `ac6` + labinfosize               | u8[width * height] | map[0]             | map data                       |
+| `ac6` + labinfosize + mapsize     | u8[width * height] | hotspots           | hotspot map                    |
+| `ac6` + labinfosize + mapsize * 2 | u8[..]             | NPC coordinates    | See below                      |
+
+where:
+* labinfosize = #labinfo * 7
+* mapsize = height * width
+
+### LabInfo blocks
+
+LabInfo blocks consist of 7 bytes each and most likely reference `LAB_DATA` and/or `LABBLOCK` resources.
+Going by similarity of the numbers observed in the various bytes, they seem to be grouped into four chunks:
+
+* One chunk of u8[4] blocks (labinfo.head, below)
+* Three chunks of u8 blocks (labinfo.rest[0], labinfo.rest[1], and labinfo.rest[2], respectively, below)
+
+| Labinfo | Meaning | Notes                                                                |
+|---------|---------|----------------------------------------------------------------------|
+| head[0] | ?       | observed values: [`00-02`, `04`, `10`]                               |
```

Comment:
If I had to guess (without looking at the labdata) I'd say the labdata defines objects and textured walls and this structure specifies which type (object, wall) it is, which automap type (wall, trap, etc) and which index (which wall/object). But as this is different from Ambermoon this is only a shot in the blue.

Reply:
I think the first 4 bytes are the tile flags. In Ambermoon they specify if the object is a floor object like a hole or liquid, for walls if it has transparent parts and also who can pass the block (wall/door/object).

Rest[0] is either the object/wall index or the automap (dungeon map) type. Rest[3] is the other one.

Rest[2] is most likely the type: object, wall and maybe free/empty block or unused (this is in Ambermoon a part which can't be seen nor moved to.

Reply:
I wasn't able to find too much that seemed similar to the Ambermoon specs in the 3D parts, but I agree that the properties you list would be a good fit for this section.

* automap: the Amberstar automap lists eight or nine different types of map entries (unless there are some secret ones that aren't shown until later, but the GUI doesn't seem set up for that), so the corresponding entry should have that many different options.  Of course the encoding might be split up in a less obvious way...  this probably corresponds to values stored in the automap files, though, so those might be a good point of comparison.
* walkable vs. not walkable might fit with `0x80` in `head[3]`, looking at the sewers and at Twinlake (assuming that the first labblock represents index `01` and not `00`).

Reply:
I guess labblock index 0 is not valid and is instead used to represent empty blocks. So the first entry is index 1 as you mentioned.

There are hidden automap types in Ambermoon. 0 is "None" as objects might have no visual type on the automap like some decoration and 1 is "Wall" which just shows the wall on the map.

Reply:
In Ambermoon the movement blocking is controlled in two ways. I guess it is the same here. 0x80 means "block all movement". Only if this is not set, the bits above control which travel type can pass.

0x100 would allow walking, 0x200 would allow horse riding etc. I see the values for the second lowest flag byte are only 0x00, 0x01, 0x03 and 0x07 which would fit. It means none of the first 8 travel types can pass (most likely used together with 0x80), only walking is allowed, walking and something else is allowed and then walking and 2 others are allowed.

I am not sure what travel types are possible in Amberstar. If there are 8 or less they should fit into the byte.

Note: In Ambermoon the swimming bit is not only used for movement checks but also to mark tiles as water.

Reply:
So all water should have a distinct bit set in those flag bytes.

Reply:
In Ambermoon the travel types are also used in 3D where you only can walk. But here more like a "collision category" where the player uses the first class "walk" and monsters often use another class. This was you can achieve walls the player can enter but the monsters not. Like fake walls or some doors.